librecrypt_settings_prefix.3
.TH LIBRECRYPT_SETTINGS_PREFIX 3 LIBRECRYPT
.SH NAME
librecrypt_settings_prefix - Get length of settings prefix in a password hash string

.SH SYNOPSIS
.nf
#include <librecrypt.h>

size_t \fBlibrecrypt_settings_prefix\fP(const char *\fIhash\fP, size_t *\fIhashsize_out\fP,
                                  LIBRECRYPT_CONTEXT *\fIctx\fP);
.fi
.PP
Link with
.IR -lrecrypt .

.SH DESCRIPTION
The
.BR librecrypt_settings_prefix ()
function returns the number of bytes, from the front of
.IR hash ,
that make up the algorithm configuration.
.PP
Unless
.I hashsize_out
is
.IR NULL ,
the value 0
is stored in
.I *hashsize_out
if the output hash size is fixed for the last algorithm in
.I hash
or if the output hash size is not specified (either because
.IR &hash[r] ,
where
.I r
is the return value, is the null byte, and so the default
hash size shall be used or because the information is
malformatted), otherwise a positive value is stored in
.IR *hashsize_out ,
specifying the number of bytes in the binary output hash,
as determined by the text in
.I hash
beyond the settings prefix.
See the
.B EXTENDED DESCRIPTION
section for more information.
.PP
The
.I ctx
parameter is used for providing library
configurations, see
.BR librecrypt_create_context (3)
for more information; used
.I NULL
for default configurations.
.PP
.I hash
must not be
.IR NULL .

.SH RETURN VALUES
The
.BR librecrypt_settings_prefix ()
function returns the number of bytes that make up the
settings prefix. The return value may be 0, which shall
be treated as any other valid return value.
.PP
.IR &hash[r] ,
where
.I r
is the return value, points to the hash result proper.

.SH ERRORS
The
.BR librecrypt_settings_prefix ()
function cannot fail.

.SH ATTRIBUTES
For an explanation of the terms used in this section, see
.BR attributes (7).
.PP
.TS
allbox;
lb lb lb
l l l.
Interface	Attribute	Value
T{
.BR librecrypt_settings_prefix ()
T}	Thread safety	MT-Safe
T{
.BR librecrypt_settings_prefix ()
T}	Async-signal safety	AS-Safe
.TE
.sp

.SH EXTENDED DESCRIPTION
Some algorithms have flexible hash size which
is encoded either with an actual hash (its
length after decoding to binary is checked)
or using asterisk-notation in place of the
result (that is, as an asterisk
.RI (\(aq * \(aq)
followed by an unsigned, decimal integer,
which may have leading zeroes). This part
is always excluded (from the last algorithm
in the algorithm chain in
.IR hash )
in the return value of the
.BR librecrypt_settings_prefix ()
function.

.SH HISTORY
The
.BR librecrypt_settings_prefix ()
function was introduced in version 1.0 of
.BR librecrypt .

Since version 1.1 of the
.BR librecrypt_settings_prefix ()
function, the
.I ctx
parameter (previously called
.IR reserved )
has the type
.B LIBRECRYPT_CONTEXT *
(previously
.BR "void *" ).

.SH SEE ALSO
.BR librecrypt (7),
.BR librecrypt_chain_length (3)


librecrypt_settings_prefix.c
/* See LICENSE file for copyright and license details. */
#include "common.h"
#ifndef TEST


size_t
librecrypt_settings_prefix(const char *hash, size_t *hashsize_out, LIBRECRYPT_CONTEXT *ctx)
{
	size_t i, len, ret = 0u;
	size_t last_offset = 0u;
	const struct librecrypt_algorithm *algo;
	uintmax_t hashsize;

	/* Find last algorithm, and beginning of result */
	for (i = 0u; hash[i]; i++) {
		if (hash[i] == LIBRECRYPT_HASH_COMPOSITION_DELIMITER)
			ret = i + 1u;
		else if (hash[i] == LIBRECRYPT_ALGORITHM_LINK_DELIMITER)
			last_offset = ret = i + 1u;
	}
	/* and get `strlen(hash)` */
	len = i;

	/* Special case for bsdicrypt */
	if (ret == last_offset && ret < len && hash[ret] == '_')
		ret += 1u;

	/* Return if hash size is not required */
	if (!hashsize_out)
		goto out;
	/* Return 0 as hash size if default is specified */
	if (ret == i)
		goto zero;
	/* Return 0 as hash size if algorithm cannot be identified or has fixed hash size */
	algo = librecrypt_find_first_algorithm_(&hash[last_offset], len - last_offset, ctx);
	if (!algo)
		goto zero;
	if (!algo->flexible_hash_size)
		goto zero;

	/* Get the hash size */
	if (!librecrypt_scan_settings(&hash[ret], len - ret, "%^b",
	                              &hashsize, (uintmax_t)1u, (uintmax_t)SIZE_MAX,
	                              algo->decoding_lut, algo->pad, algo->strict_pad))
		goto zero;

	*hashsize_out = (size_t)hashsize;
	goto out;

zero:
	*hashsize_out = 0u;
out:
	return ret;
}


#else
# ifndef FUZZ


#define CHECK_NULL(PREFIX, SUFFIX)\
	do {\
		EXPECT(librecrypt_settings_prefix(PREFIX SUFFIX, NULL, ctx) == sizeof(PREFIX) - 1u);\
		EXPECT(librecrypt_settings_prefix(PREFIX, NULL, ctx) == sizeof(PREFIX) - 1u); \
	} while (0)

#define CHECK_ZERO(PREFIX, SUFFIX)\
	do {\
		size_t hashsize = 99999u;\
		EXPECT(librecrypt_settings_prefix(PREFIX SUFFIX, &hashsize, ctx) == sizeof(PREFIX) - 1u);\
		EXPECT(hashsize == 0u);\
	} while (0)

#define CHECK_HASH(PREFIX, SUFFIX, HASH)\
	do {\
		size_t hashsize = 99999u;\
		EXPECT(librecrypt_settings_prefix(PREFIX SUFFIX, &hashsize, ctx) == sizeof(PREFIX) - 1u);\
		EXPECT(hashsize == HASH##u);\
	} while (0)


static unsigned
dumdum_is_algorithm(const char *settings, size_t len)
{
	(void) settings;
	(void) len;
	return 1u;
}


static struct librecrypt_algorithm dumdum = {
	.is_algorithm = &dumdum_is_algorithm,
	.flexible_hash_size = 0
};


int
main(void)
{
	LIBRECRYPT_CONTEXT *ctx = NULL;

	SET_UP_ALARM();
	INIT_RESOURCE_TEST();

	/* Simple cases */
	CHECK_NULL("", "result");
	CHECK_NULL(">", "double-des result");
	CHECK_NULL("$", "x");
	CHECK_NULL("y$", "x");
	CHECK_NULL("y>", "x");
	CHECK_NULL("a$b$c>d$e$", "x");
	CHECK_NULL("a$b$c>", "x");

	/* Discard asterisk-notation */
	CHECK_NULL("", "*10");
	CHECK_NULL(">", "*11");
	CHECK_NULL("$", "*12");
	CHECK_NULL("y$", "*13");
	CHECK_NULL("y>", "*012");
	CHECK_NULL("a$b$c>d$e$", "*1");
	CHECK_NULL("a$b$c>", "*2");
	CHECK_NULL("a$b$c>*10$", "*2");

	/* Special case: bsdicrypt */
	CHECK_NULL("_", "res");
	CHECK_NULL("_", "");
	CHECK_NULL("$x$*10>_", "_");

	/* Check without hash */
	CHECK_ZERO("a$b$", "");
	CHECK_ZERO("a$b$>", "");
	CHECK_ZERO("a$b$x>", "");
	CHECK_ZERO("a$b$*10>", "");

	/* Check without invalid algorithm */
	CHECK_ZERO("$~no~such~algorithm~$", "hash");

	/* Check without hash and hashlen */
#if defined(SUPPORT_ARGON2I)
	CHECK_HASH("$argon2i$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("x$*99>$argon2i$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("$argon2i$m=8,t=1,p=1$*99$", "NineByteHash", 9);
	CHECK_HASH("x$*99>$argon2i$m=8,t=1,p=1$*99$", "NineByteHash", 9);
#endif
#if defined(SUPPORT_ARGON2ID)
	CHECK_HASH("$argon2id$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("x$*99>$argon2id$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("$argon2id$m=8,t=1,p=1$*99$", "NineByteHash", 9);
	CHECK_HASH("x$*99>$argon2id$m=8,t=1,p=1$*99$", "NineByteHash", 9);
#endif
#if defined(SUPPORT_ARGON2D)
	CHECK_HASH("$argon2d$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("x$*99>$argon2d$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("$argon2d$m=8,t=1,p=1$*99$", "NineByteHash", 9);
	CHECK_HASH("x$*99>$argon2d$m=8,t=1,p=1$*99$", "NineByteHash", 9);
#endif
#if defined(SUPPORT_ARGON2DS)
	CHECK_HASH("$argon2ds$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("x$*99>$argon2ds$m=8,t=1,p=1$*99$", "*100", 100);
	CHECK_HASH("$argon2ds$m=8,t=1,p=1$*99$", "NineByteHash", 9);
	CHECK_HASH("x$*99>$argon2ds$m=8,t=1,p=1$*99$", "NineByteHash", 9);
#endif

	/* Check without invalid hash */
	IF__argon2i__SUPPORTED(CHECK_ZERO("$argon2i$m=8,t=1,p=1$*99$", "#");)
	IF__argon2d__SUPPORTED(CHECK_ZERO("$argon2d$m=8,t=1,p=1$*99$", "#");)
	IF__argon2id__SUPPORTED(CHECK_ZERO("$argon2id$m=8,t=1,p=1$*99$", "#");)
	IF__argon2ds__SUPPORTED(CHECK_ZERO("$argon2ds$m=8,t=1,p=1$*99$", "#");)

	/* Test without flexible hash size (but with context I guess) */
	ctx = librecrypt_create_context();
	assert(ctx != NULL);
	librecrypt_set_custom_algorithms(ctx, &dumdum, 1u);
	CHECK_ZERO("$dumdum$", "abcd");
	librecrypt_free_context(ctx);

	STOP_RESOURCE_TEST();
	return 0;
}


# else


int
LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
{
	char *hash;
	size_t r;
	hash = malloc(size + 1u);
	assert(hash);
	memcpy(hash, data, size);
	hash[size] = '\0';
	r = librecrypt_settings_prefix(hash, &(size_t){0u}, NULL);
	EXPECT(librecrypt_settings_prefix(hash, NULL, NULL) == r);
	free(hash);
	return 0;
}


# endif
#endif
